### ISSUE TITLE
Is the rendered flow a backward flow?

### ISSUE THREAD

**Issue by iimmortall** [openxrlab/xrfeitoria#38]
Excellent work! Is the rendered flow a backward flow?

**Comment by HaiyiMei** [comment]
Thanks for your attention.
You mean is the rendering process differentiable? If this is the question, then the answer is no. Cos the backends of our project are blender and unreal engine, which are not differentiable.

**Comment by iimmortall** [comment]
Sorry for the confusion. I’m working on image warping using rendered optical flow, but I’m uncertain whether the rendered optical flow represents the motion from the current frame to the next frame (forward) or from the current frame to the previous frame (backward). Could you clarify?

**Comment by HaiyiMei** [comment]
I think for forward, yes. The `flow` in [here](https://xrfeitoria.readthedocs.io/en/latest/apis/generated/xrfeitoria.data_structure.models.RenderPass.html) represents the forward optical flow. 
But as for backward, our package doesn't support such features unfortunately. I'll recommend you to use [BlenderProc](https://dlr-rm.github.io/BlenderProc/examples/advanced/optical_flow/README.html).

**Comment by HaiyiMei** [comment]
Yes, it is both possible for forward and backward flow. But I just discovered there is a bug in backward flow. I'll notify you once it's fixed. Before that, you can refer to this [BlenderProc](https://github.com/DLR-RM/BlenderProc/blob/85c19ada83e2cad81866de4d7734a70767ddf017/blenderproc/python/renderer/FlowRendererUtility.py#L136-L145).

**Comment by iimmortall** [comment]
Thank you, I am looking forward to your good news. Is the optical flow in the current version forward optical flow? And is the rendered optical flow accurate?

**Comment by HaiyiMei** [comment]
I fixed it in this commit https://github.com/openxrlab/xrfeitoria/commit/8f8f2e041fbc4c84b78ad30a70fd9f3eba2f8248. You can use it by following these steps for now (before a new version is released):

```bash
git clone https://github.com/openxrlab/xrfeitoria.git
cd xrfeitoria
git checkout alpha_channel
pip install .
python -c "import xrfeitoria as xf; xf.init_blender(replace_plugin=True, dev_plugin=True)"
```


> Thank you, I am looking forward to your good news. Is the optical flow in the current version forward optical flow? And is the rendered optical flow accurate?

As for your question, yes I think now the forward and backward flow are all accurate.

And as for how to read `.exr` files, you can refer to our code [here](https://github.com/openxrlab/xrprimer/blob/main/python/xrprimer/io/exr_reader.py).

**Comment by iimmortall** [comment]
I use UE5 as my dev, I find the render optical flow just `rg` channel has value,  does that mean it is forward optical flow? And is the rendered optical flow accurate at UE5 dev?

**Comment by HaiyiMei** [comment]
> I use UE5 as my dev, I find the render optical flow just `rg` channel has value, does that mean it is forward optical flow? And is the rendered optical flow accurate at UE5 dev?

Yes, in ue5 `rg` is the forward optical flow, and you should also export it as `exr` file when rendering. I think it's pretty accurate. Unfortunately, I'm unsure if it's possible to easily generate backward flow.

**Comment by iimmortall** [comment]
Thank you! If there are any plans to support generating backward optical flow with UE5, please let me know.

**Comment by iimmortall** [comment]
Hello, I noticed that the all values of the rendered optical flow (EXR) are quite small, typically less than 1.0. How should these values be properly scaled? For instance, should they be scaled based on the rendered image size, such as flow[:,:,0] * width and flow[:,:,1] * height?

**Comment by HaiyiMei** [comment]
Hi, firstly, you can try to use this [class](https://github.com/openxrlab/xrfeitoria/blob/main/xrfeitoria/utils/viewer.py#L101) to visualize the flow to see if it's weird.

**Comment by iimmortall** [comment]
The visualized optical flow appears correct, as the [flow_vis.flow_to_color() ](https://github.com/tomrunia/OpticalFlow_Visualization/blob/master/flow_vis/flow_vis.py) function scales the flow appropriately when converting it to an RGB image.

**Comment by HaiyiMei** [comment]
I guess it's because the object is too far away from the camera? The optical flow is in camera space, so the value is tiny. Try to place the object near the camera, and see if the value is getting significantly bigger?

**Comment by Langwenchong** [comment]
@iimmortall  hello，have you solved this issue? I also found that rendered flow  exr file by unreal5.1 only channel r/g has value and it is too small only around 1e-6

**Comment by Langwenchong** [comment]
@HaiyiMei Hello, I would like to consult you about the exact meaning of the values and directions of optical flow (forward optical flow).

Based on my experimental comparisons, I suspect that the optical flow rendered by UE5.1 uses a coordinate system where the origin is at the bottom-left corner, with the positive x-axis pointing to the right and the positive y-axis pointing upward. This differs from the image coordinate system commonly used in OpenCV or NumPy, where the origin is at the top-left corner, the positive x-axis points to the right, and the positive y-axis points downward.

As a result, I believe that in the exported optical flow EXR file, the G channel (which represents the v component, i.e., the y-direction) needs to be multiplied by −1 to align with the OpenCV image coordinate convention.

In addition, I would like to ask whether the u and v values of the optical flow rendered by UE5.1 via xrfeitoria need to be scaled by W/2 and H/2 respectively.

I urgently need to resolve this issue. I would greatly appreciate it if you could reply when you have time. Thank you very much.
